# Question 2: Numerical Analysis Using NumPy

## 1. Introduction

This section performs numerical analysis on South African GDP growth and employment data using Python and NumPy.

The dataset contains observations for South Africa across different years and economic sectors. NumPy is used to perform numerical calculations, descriptive statistics, array operations, filtering, reshaping, correlation analysis, and identification of extreme observations.

The purpose of the analysis is to identify important numerical patterns in GDP growth and employment and demonstrate how NumPy can be used to analyse economic data efficiently.

In [28]:
import pandas as pd
import numpy as np

In [29]:
try:
    df = pd.read_csv("../../data/final_clean/GDP_Employment_South_Africa_Cleaned.csv")
except FileNotFoundError:
    print("Cleaned dataset not found. Please run the data preparation step first.")
    raise

df.head()

,REF_AREA_LABEL,TIME_PERIOD,GDP_Growth,Employment_Thousands,Sector
0,South Africa,2017,1.157947,16536.302734,Sector: Total
1,South Africa,2017,1.157947,849.620773,Sector: Agriculture
2,South Africa,2017,1.157947,414.533398,Sector: Mining
3,South Africa,2017,1.157947,1855.818421,Sector: Manufacturing
4,South Africa,2017,1.157947,187.111598,Sector: Utilities


In [30]:
print("Number of rows:", len(df))
print("Number of columns:", len(df.columns))
print("Columns:")
print(df.columns.tolist())

print("\nCountries:")
print(df["REF_AREA_LABEL"].unique())

print("\nNumber of unique countries:", df["REF_AREA_LABEL"].nunique())

Number of rows: 570
Number of columns: 5
Columns:
['REF_AREA_LABEL', 'TIME_PERIOD', 'GDP_Growth', 'Employment_Thousands', 'Sector']

Countries:
['South Africa']

Number of unique countries: 1


## 2. Data Quality Verification

Before performing numerical analysis, the dataset is checked for correct data types, missing values, and duplicate observations. This ensures that the NumPy calculations are performed on valid numerical data.

In [31]:
print("Data types:")
print(df.dtypes)

print("\nMissing values:")
print(df.isnull().sum())

print("\nDuplicate rows:")
print(df.duplicated().sum())

Data types:
REF_AREA_LABEL           object
TIME_PERIOD               int64
GDP_Growth              float64
Employment_Thousands    float64
Sector                   object
dtype: object

Missing values:
REF_AREA_LABEL          0
TIME_PERIOD             0
GDP_Growth              0
Employment_Thousands    0
Sector                  0
dtype: int64

Duplicate rows:
0


## 3. NumPy Array Conversion

The numerical variables are converted from Pandas Series into NumPy arrays. This allows NumPy functions to be used for statistical calculations and numerical operations.

In [32]:
years = df["TIME_PERIOD"].to_numpy()
gdp_growth = df["GDP_Growth"].to_numpy()
employment = df["Employment_Thousands"].to_numpy()

print("Years array shape:", years.shape)
print("GDP Growth array shape:", gdp_growth.shape)
print("Employment array shape:", employment.shape)

print("\nFirst 10 years:")
print(years[:10])

print("\nFirst 10 GDP Growth values:")
print(gdp_growth[:10])

print("\nFirst 10 Employment values:")
print(employment[:10])

Years array shape: (570,)
GDP Growth array shape: (570,)
Employment array shape: (570,)

First 10 years:
[2017 2017 2017 2017 2017 2017 2017 2017 2017 2017]

First 10 GDP Growth values:
[1.15794695 1.15794695 1.15794695 1.15794695 1.15794695 1.15794695
 1.15794695 1.15794695 1.15794695 1.15794695]

First 10 Employment values:
[16536.30273438   849.62077318   414.53339754  1855.81842055
   187.11159817  1381.0282351   3307.47734866  1026.52085431
  2495.86583723  5018.3255011 ]


Note: GDP growth is a national yearly figure, not specific to each sector. It repeats across sector rows because it was joined to sector-level employment data — this is why the same GDP growth value appears multiple times for a given year.

## 4. Descriptive Statistics

NumPy statistical functions are used to calculate the mean, median, minimum, maximum, standard deviation, and variance for GDP growth and employment.

These measures provide an overview of the central tendency and variation in the South African economic data.

In [33]:
# GDP Growth statistics
gdp_mean = np.mean(gdp_growth)
gdp_median = np.median(gdp_growth)
gdp_min = np.min(gdp_growth)
gdp_max = np.max(gdp_growth)
gdp_std = np.std(gdp_growth)
gdp_variance = np.var(gdp_growth)

print("GDP Growth Mean:", gdp_mean)
print("GDP Growth Median:", gdp_median)
print("GDP Growth Minimum:", gdp_min)
print("GDP Growth Maximum:", gdp_max)
print("GDP Growth Standard Deviation:", gdp_std)
print("GDP Growth Variance:", gdp_variance)

print("\n" + "="*60 + "\n")

# Employment statistics
employment_mean = np.mean(employment)
employment_median = np.median(employment)
employment_min = np.min(employment)
employment_max = np.max(employment)
employment_std = np.std(employment)
employment_variance = np.var(employment)

print("Employment Mean:", employment_mean)
print("Employment Median:", employment_median)
print("Employment Minimum:", employment_min)
print("Employment Maximum:", employment_max)
print("Employment Standard Deviation:", employment_std)
print("Employment Variance:", employment_variance)

GDP Growth Mean: 3.023583474342852
GDP Growth Median: 3.1000000005441137
GDP Growth Minimum: -2.1370328450052085
GDP Growth Maximum: 7.939608644806768
GDP Growth Standard Deviation: 2.409849529155861
GDP Growth Variance: 5.8073747531727244


Employment Mean: 1870.569163593318
Employment Median: 947.9243344425784
Employment Minimum: 17.798834023718744
Employment Maximum: 16536.302734375
Employment Standard Deviation: 2892.597946701926
Employment Variance: 8367122.881264198


Employment's mean is much higher than its median. This happens because a few large sectors employ far more people than most sectors, pulling the average up.

## 5. Range and Percentile Analysis

The range measures the difference between the maximum and minimum values. NumPy's `np.ptp()` function is used to calculate the range.

The 25th, 50th, and 75th percentiles are also calculated to describe the distribution of GDP growth and employment.

In [34]:
# GDP Growth range and percentiles
gdp_range = np.ptp(gdp_growth)
gdp_percentiles = np.percentile(gdp_growth, [25, 50, 75])

print("GDP Growth Range:", gdp_range)
print("GDP Growth 25th, 50th and 75th Percentiles:", gdp_percentiles)

print("\n" + "="*60 + "\n")

# Employment range and percentiles
employment_range = np.ptp(employment)
employment_percentiles = np.percentile(employment, [25, 50, 75])

print("Employment Range:", employment_range)
print("Employment 25th, 50th and 75th Percentiles:", employment_percentiles)

GDP Growth Range: 10.076641489811976
GDP Growth 25th, 50th and 75th Percentiles: [1.41382645 3.1        4.57194475]


Employment Range: 16518.503900351283
Employment 25th, 50th and 75th Percentiles: [ 402.8830962   947.92433444 1779.99505316]


## 6. NumPy Array Reshaping

NumPy arrays can be reshaped into different dimensions without changing their underlying values. Both GDP growth and employment arrays are reshaped from one-dimensional arrays into two-dimensional column arrays.

In [35]:
# Reshape both numerical arrays into column vectors
gdp_reshaped = gdp_growth.reshape(-1, 1)
employment_reshaped = employment.reshape(-1, 1)

print("Original GDP Growth shape:", gdp_growth.shape)
print("Reshaped GDP Growth shape:", gdp_reshaped.shape)

print("\nOriginal Employment shape:", employment.shape)
print("Reshaped Employment shape:", employment_reshaped.shape)

print("\nFirst 5 GDP Growth values after reshaping:")
print(gdp_reshaped[:5])

print("\nFirst 5 Employment values after reshaping:")
print(employment_reshaped[:5])

Original GDP Growth shape: (570,)
Reshaped GDP Growth shape: (570, 1)

Original Employment shape: (570,)
Reshaped Employment shape: (570, 1)

First 5 GDP Growth values after reshaping:
[[1.15794695]
 [1.15794695]
 [1.15794695]
 [1.15794695]
 [1.15794695]]

First 5 Employment values after reshaping:
[[16536.30273438]
 [  849.62077318]
 [  414.53339754]
 [ 1855.81842055]
 [  187.11159817]]


## 7. GDP Growth and Employment Correlation

The Pearson correlation coefficient is calculated using NumPy to measure the linear relationship between GDP growth and employment.

The correlation coefficient ranges from -1 to +1. Values close to +1 indicate a strong positive relationship, values close to -1 indicate a strong negative relationship, and values close to 0 indicate a weak or no linear relationship.

In [36]:

# GDP GROWTH AND TOTAL EMPLOYMENT CORRELATION


# GDP growth is recorded once per year.
# Use only the "Sector: Total" employment values so that
# GDP growth and employment have one observation per year.

total_mask = df["Sector"].to_numpy() == "Sector: Total"

annual_gdp = gdp_growth[total_mask]
annual_employment = employment[total_mask]
annual_years = years[total_mask]

# Check that we have one observation per year
print("Number of yearly observations:", len(annual_years))
print("First year:", annual_years.min())
print("Last year:", annual_years.max())

# Calculate correlation using NumPy
correlation = np.corrcoef(annual_gdp, annual_employment)[0, 1]

print("\n===== ANNUAL GDP GROWTH AND TOTAL EMPLOYMENT CORRELATION =====")
print(f"Correlation: {correlation:.4f}")

Number of yearly observations: 57
First year: 1961
Last year: 2017

===== ANNUAL GDP GROWTH AND TOTAL EMPLOYMENT CORRELATION =====
Correlation: -0.2975


The annual correlation between GDP growth and total employment was approximately -0.30. This indicates a weak negative relationship between the two variables during the period analysed. However, correlation does not mean that GDP growth directly causes employment to increase or decrease. The result suggests that annual GDP growth and employment did not consistently move in the same direction.


## 8. Sector-Level Numerical Analysis

The dataset contains employment observations for different economic sectors in South Africa. NumPy is used to calculate employment statistics for each economic sector. Since GDP growth is a national yearly measure rather than a sector-specific measure, the sector comparison focuses primarily on employment differences


This allows differences in employment levels between sectors to be identified numerically.

In [37]:
# Create a NumPy-based sector summary

sector_results = []
sectors = df[df["Sector"] != "Sector: Total"]["Sector"].unique()

for sector in sectors:
    sector_mask = df["Sector"].to_numpy() == sector
    
    sector_gdp = gdp_growth[sector_mask]
    sector_employment = employment[sector_mask]
    
    sector_results.append([
        sector,
        np.mean(sector_gdp),
        np.mean(sector_employment),
        np.min(sector_employment),
        np.max(sector_employment)
    ])

sector_results = np.array(sector_results, dtype=object)

print("Sector | Mean GDP Growth | Mean Employment | Min Employment | Max Employment")

for row in sector_results:
    print(
        f"{row[0]} | "
        f"{row[1]:.2f} | "
        f"{row[2]:.2f} | "
        f"{row[3]:.2f} | "
        f"{row[4]:.2f}"
    )

Sector | Mean GDP Growth | Mean Employment | Min Employment | Max Employment
Sector: Agriculture | 3.02 | 1039.69 | 661.44 | 1409.83
Sector: Mining | 3.02 | 473.72 | 334.17 | 657.36
Sector: Manufacturing | 3.02 | 1473.76 | 653.89 | 2159.71
Sector: Utilities | 3.02 | 72.42 | 17.80 | 187.11
Sector: Construction | 3.02 | 642.53 | 221.54 | 1409.38
Sector: Trade services | 3.02 | 2071.34 | 818.11 | 3682.75
Sector: Transport services | 3.02 | 515.61 | 189.06 | 1026.52
Sector: Finance and business services | 3.02 | 831.60 | 136.60 | 2495.87
Sector: Other services | 3.02 | 2232.17 | 810.84 | 5018.33


Other Services has the highest average employment (2,232.17 thousand), followed closely by Trade Services (2,071.34 thousand). Utilities has by far the lowest average employment (72.42 thousand), followed by Mining (473.72 thousand). This shows that employment in South Africa is concentrated mainly in services-related sectors, while Utilities and Mining employ far fewer people on average.

## 9. Identifying Minimum and Maximum Observations

NumPy's `argmin()` and `argmax()` functions are used to identify the positions of the minimum and maximum values in the GDP Growth and Employment arrays.

The corresponding rows in the dataset are then used to determine the year and sector associated with these extreme values.

In [38]:
# Find indexes of minimum and maximum GDP Growth
gdp_min_index = np.argmin(gdp_growth)
gdp_max_index = np.argmax(gdp_growth)

# Find indexes of minimum and maximum Employment
employment_min_index = np.argmin(employment)
employment_max_index = np.argmax(employment)

print("GDP Growth Minimum:")
print("Value:", gdp_growth[gdp_min_index])
print("Year:", df.iloc[gdp_min_index]["TIME_PERIOD"])
print("Sector:", df.iloc[gdp_min_index]["Sector"])

print("\nGDP Growth Maximum:")
print("Value:", gdp_growth[gdp_max_index])
print("Year:", df.iloc[gdp_max_index]["TIME_PERIOD"])
print("Sector:", df.iloc[gdp_max_index]["Sector"])

print("\nEmployment Minimum:")
print("Value:", employment[employment_min_index])
print("Year:", df.iloc[employment_min_index]["TIME_PERIOD"])
print("Sector:", df.iloc[employment_min_index]["Sector"])

print("\nEmployment Maximum:")
print("Value:", employment[employment_max_index])
print("Year:", df.iloc[employment_max_index]["TIME_PERIOD"])
print("Sector:", df.iloc[employment_max_index]["Sector"])

GDP Growth Minimum:
Value: -2.1370328450052085
Year: 1992
Sector: Sector: Total

GDP Growth Maximum:
Value: 7.939608644806768
Year: 1964
Sector: Sector: Total

Employment Minimum:
Value: 17.798834023718744
Year: 1961
Sector: Sector: Utilities

Employment Maximum:
Value: 16536.302734375
Year: 2017
Sector: Sector: Total


The lowest GDP growth (-2.14%) occurred in 1992, while the highest GDP growth (7.94%) occurred in 1964 — both at the national (Total) level. The lowest single employment figure (17.80 thousand) was in the Utilities sector in 1961, while the highest employment figure (16,536.30 thousand) was the national Total in 2017. This shows that employment naturally grew over time as the population and economy expanded, while GDP growth has fluctuated up and down across different years.

## 10. Sorting and Ranking Employment Values

NumPy sorting functions are used to rank employment values from lowest to highest. This helps identify the smallest and largest employment observations in the dataset.

In [39]:
# Sort employment values

sorted_employment = np.sort(employment)

print("Lowest 10 employment values:")
print(sorted_employment[:10])

print("\nHighest 10 employment values:")
print(sorted_employment[-10:])

Lowest 10 employment values:
[17.79883402 18.26115301 19.78919128 20.67919445 21.60922286 23.38754948
 24.0180168  26.85948457 27.6073459  29.32977994]

Highest 10 employment values:
[14150.99609375 14377.88085938 14595.64648438 14709.78515625
 15076.05371094 15159.45019531 15415.94238281 16130.1015625
 16188.35253906 16536.30273438]


In [40]:

# GDP GROWTH DIRECTION BY YEAR

positive_years = np.sum(annual_gdp > 0)
negative_years = np.sum(annual_gdp < 0)
zero_years = np.sum(annual_gdp == 0)

print("===== GDP GROWTH DIRECTION BY YEAR =====")
print("Positive-growth years:", positive_years)
print("Negative-growth years:", negative_years)
print("Zero-growth years:", zero_years)
print("Total years:", len(annual_gdp))

===== GDP GROWTH DIRECTION BY YEAR =====
Positive-growth years: 49
Negative-growth years: 8
Zero-growth years: 0
Total years: 57


The lowest employment values (around 17 to 29 thousand) come from smaller categories like Utilities in earlier years, while the highest values (around 14,000 to 16,536 thousand) are mostly the yearly national Totals from more recent years. This confirms that employment figures include both individual sectors and the combined national total, which explains the very wide gap between the smallest and largest numbers.

## Summary of Findings:

GDP growth in South Africa averaged 3.02% over the period studied, ranging from -2.14% (1992) to 7.94% (1964). Of the 57 years analysed, 49 recorded positive GDP growth while 8 recorded negative growth. Employment was far more unevenly distributed, with its mean (1,870.57 thousand) being almost double its median (947.92 thousand). This difference was driven mainly by sectors such as Other Services and Trade Services, which employed substantially more people than smaller sectors such as Utilities and Mining.

The annual correlation between GDP growth and total employment was -0.2975, indicating a weak negative relationship between the two variables. This means that higher GDP growth did not consistently correspond with higher total employment during the period studied. The finding suggests that employment may also be influenced by longer-term sector and structural factors rather than short-term economic growth alone. This provides a useful insight for the group's overall analysis of the relationship between economic performance and employment in South Africa.
